espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ViewFinder.java

```diff
@@ -0,0 +1,144 @@
+    }
+
+    ///Find By different strategies
+    private static List<ViewInteraction> findAllBy(Strategy strategy, String selector, Boolean findOne) throws InvalidStrategyException, XPathLookupException {
+        List<ViewInteraction> matcher;
+
+        switch (strategy) {
+            case ID: // with ID
+
+                // find id from target context
+                Context context = InstrumentationRegistry.getTargetContext();
+                int id = context.getResources().getIdentifier(selector, "Id",
+                        InstrumentationRegistry.getTargetContext().getPackageName());
+
+                matcher = getViewInteractions(withId(id), findOne);
+                break;
+            case CLASS_NAME:
+                // with class name
+                // TODO: improve this finder with instanceOf
+                matcher = getViewInteractions(withClassName(endsWith(selector)), findOne);
+                break;
+            case TEXT:
+                // with text
+                matcher = getViewInteractions(withText(selector), findOne);
+                break;
+            case ACCESSIBILITY_ID:
+                // with content description
+                matcher = getViewInteractions(withContentDescription(selector), findOne);
+                break;
+            case XPATH:
+                // If we're only looking for one item that matches xpath, pass it index 0 or else
+                // Espresso throws an AmbiguousMatcherException
+                if (findOne) {
+                    matcher = getViewInteractions(withXPath(selector, 0), true);
+                } else {
+                    matcher = getViewInteractions(withXPath(selector), false);
+                }
+                break;
+            default:
+                throw new InvalidStrategyException(String.format("Strategy is not implemented: %s", strategy.getStrategyName()));
+        }
+
+        return matcher;
+    }
+
+    private static List<ViewInteraction> getViewInteractions(Matcher<View> matcher, boolean findOne) {
+        if (findOne) {
+            // If it's just one view we want, return a singleton list
+            return Collections.singletonList(onView(matcher));
+        } else {
+            // If we want all views that match the criteria, start looking for ViewInteractions by
+            // index and add each match to the List. As soon as we find no match, break the loop
+            // and return the list
+            List<ViewInteraction> viewInteractions = new ArrayList<>();
+            int i = 0;
+            do {
+                ViewInteraction viewInteraction = onView(withIndex(matcher, i));
+                try {
+                    viewInteraction.check(matches(isDisplayed()));
+                } catch (NoMatchingViewException nme) {
+                    break;
+                }
+                viewInteractions.add(viewInteraction);
+                i++;
+            } while (true);
```

Comment:
can one ever have the situation that this loop cannot exit? I'd rather change this condition to i < Integer.MAX_VALUE, just for safety reasons.